libcloud/test/test_connection.py

```diff
@@ -278,5 +280,49 @@ def test_log_curl(self):
         cmd = con._log_curl(method='HEAD', url=url, body=body, headers=headers)
         self.assertEqual(cmd, 'curl -i --head --compress http://example.com:80/test/path')
 
+    def _raise_socket_error(self):
+        raise socket.gaierror('')
+
+    @retry(sleep=2)
+    def _retry_test_with_sleep(self):
+        return self._raise_socket_error()
+
+    @retry(backoff=2)
+    def _retry_test_with_backoff(self):
+        return self._raise_socket_error()
+
+    @retry(timeout=2)
+    def _retry_test_with_timeout(self):
+        return self._raise_socket_error()
+
+    def test_retry_connection(self):
+        con = Connection()
+        con.connection = Mock()
```

Comment:
As discussed IRL, something more like (syntax might be slightly off, going off memory):

```
with patch('libcloud.common.base.Connection.connect') as patched_connect:
     # do the tests
    self.assertGreaterThan(patched_connect.call_count, 1)
```